lib/arpv4/arpv4.ml

```diff
@@ -122,27 +123,30 @@ module Make (Ethif : V1_LWT.ETHIF) (Clock : V1.CLOCK) (Time : V1_LWT.TIME) = str
           let tha = arp.sha in
           let spa = arp.tpa in (* the requested address *)
           let tpa = arp.spa in (* the requesting host IPv4 *)
-          output t (Arpv4_wire.{ op=Reply; sha; tha; spa; tpa })
+          output t ~source:sha ~destination:tha (Arpv4_wire.{ op=Reply; sha; tha; spa; tpa })
 
   (* Send a gratuitous ARP for our IP addresses *)
   let output_garp t =
-    let tha = Macaddr.broadcast in
     let sha = Ethif.mac t.ethif in
+    let tha = sha in
     let tpa = Ipaddr.V4.any in
     Lwt_list.iter_s (fun spa ->
-        Log.info (fun f -> f "ARP: sending gratuitous from %a" Ipaddr.V4.pp_hum spa);
-        output t Arpv4_packet.({ op=Arpv4_wire.Reply; tha; sha; tpa; spa })
+        let arp = Arpv4_packet.({ op=Arpv4_wire.Request; tha; sha; tpa; spa }) in
```

Comment:
If I'm reading the spec correctly, the GARP should be an ARP request (which is good here) where `spa = tpa = your ip` (here `tpa` is still `Ipaddr.V4.any`).  The `tha` is not defined (common stacks use `00:00:00:00:00:00` for sending, and ignore it on receiving (`tcpdump` prints tha if it is not all zeroes).